packages/notebook/src/history.ts

```diff
@@ -0,0 +1,420 @@
+    this.translator = options.translator || nullTranslator;
+    void this._handleKernel().then(() => {
+      this.sessionContext.kernelChanged.connect(this._handleKernel, this);
+    });
+    this.toRequest = this.requestBatchSize;
+  }
+
+  /**
+   * The client session used by the foreign handler.
+   */
+  readonly sessionContext: ISessionContext;
+
+  /**
+   * Translator to be used for warnings
+   */
+  readonly translator: ITranslator;
+
+  /**
+   * The number of history items to request.
+   */
+  private toRequest: number;
+
+  /**
+   * The number of history items to increase a batch size by per subsequent request.
+   */
+  private requestBatchSize: number = 10;
+
+  /**
+   * The current editor used by the history manager.
+   */
+  get editor(): CodeEditor.IEditor | null {
+    return this._editor;
+  }
+
+  set editor(value: CodeEditor.IEditor | null) {
+    if (this._editor === value) {
+      return;
+    }
+
+    const prev = this._editor;
+    if (prev) {
+      prev.model.sharedModel.changed.disconnect(this.onTextChange, this);
+    }
+
+    this._editor = value;
+
+    if (value) {
+      value.model.sharedModel.changed.connect(this.onTextChange, this);
+    }
+  }
+
+  /**
+   * The placeholder text that a history session began with.
+   */
+  get placeholder(): string {
+    return this._placeholder;
+  }
+
+  /**
+   * Kernel session number for filtering
+   */
+  get kernelSession(): string {
+    return this._kernelSession;
+  }
+
+  /**
+   * Get whether the console history manager is disposed.
+   */
+  get isDisposed(): boolean {
+    return this._isDisposed;
+  }
+
+  /**
+   * Dispose of the resources held by the console history manager.
+   */
+  dispose(): void {
+    this._isDisposed = true;
+    this._history.length = 0;
+    Signal.clearData(this);
+  }
+
+  /**
+   * Set placeholder and editor. Start session if one is not already started.
+   *
+   * @param activeCell - The currently selected Cell in the notebook.
+   */
+  protected async checkSession(activeCell: Cell): Promise<void> {
+    if (!this._hasSession) {
+      await this._retrieveHistory();
+      this._hasSession = true;
+      this.editor = activeCell.editor;
+      this._placeholder = this._editor?.model.sharedModel.getSource() || '';
+      // Filter the history with the placeholder string.
+      this.setFilter(this._placeholder);
+      this._cursor = this._filtered.length - 1;
+    }
+  }
+
+  /**
+   * Get the previous item in the console history.
+   *
+   * @param activeCell - The currently selected Cell in the notebook.
+   *
+   * @returns A Promise for console command text or `undefined` if unavailable.
+   */
+  async back(activeCell: Cell): Promise<string> {
+    await this.checkSession(activeCell);
+    --this._cursor;
+    if (this._cursor < 0) {
+      await this.fetchBatch();
+    }
+    this._cursor = Math.max(0, this._cursor);
+    const content = this._filtered[this._cursor];
+    return content;
```

Comment:
The typing suggests that it will never return undefined as it will stop at the oldest entry. But what happens if there is no history yet? In a very fresh installation, `this._filtered[0]` would resolve to `undefined`, right? Will the downstream error out due to incorrect typing? I would suggest updating the return type to `Promise<string | undefined>` and making any changes that may be needed in the calling functions accordingly.

Reply:
The source of the active cell is added to `this._filtered` in `setFilter` and it isn't possible to access `_filtered` before `setFilter` is run so it shouldn't be possible. I can update it just in case though.

Reply:
In that case no need to change code, but adding a comment would be helpful.

Reply:
Or ideally a unit test case